renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -2137,13 +2157,37 @@ struct D3D12PixelHistoryPerFragmentCallback : D3D12PixelHistoryCallback
     D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC pipeDesc;
     origPSO->Fill(pipeDesc);
 
-    // TODO: We provide our own render target here, but some of the PSOs use the original shader,
-    // which may write to a different number of channels (such as a R11G11B10 render target).
-    // This results in a D3D12 debug layer warning, but shouldn't be hazardous in practice since
-    // we restrict what we read from based on the source RT target. But maybe there's a way to
-    // gather the right data and avoid  the debug layer warning.
-    pipeDesc.RTVFormats.NumRenderTargets = 1;
-    pipeDesc.RTVFormats.RTFormats[0] = DXGI_FORMAT_R32G32B32A32_FLOAT;
+    if(colorOutputIndex != D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT)
+    {
+      // TODO: We provide our own render target here, but some of the PSOs use the original shader,
+      // which may write to a different number of channels (such as a R11G11B10 render target).
+      // This results in a D3D12 debug layer warning, but shouldn't be hazardous in practice since
+      // we restrict what we read from based on the source RT target. But maybe there's a way to
+      // gather the right data and avoid  the debug layer warning.
+      DXGI_FORMAT colorFormat = DXGI_FORMAT_R32G32B32A32_FLOAT;
+      // For this pass, we will be binding the color target for capturing the shader output to the
+      //  slot in question. The others can be left intact.
+      RDCASSERT(colorOutputIndex < pipeDesc.RTVFormats.NumRenderTargets);
+      pipeDesc.RTVFormats.RTFormats[colorOutputIndex] = colorFormat;
+
+      // In order to have different write masks per render target, we need to switch to independent
+      //  blend if not already in use.
+      if(!pipeDesc.BlendState.IndependentBlendEnable)
+      {
+        pipeDesc.BlendState.IndependentBlendEnable = TRUE;
+        for(uint32_t i = 1; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; ++i)
+          pipeDesc.BlendState.RenderTarget[i] = pipeDesc.BlendState.RenderTarget[0];
+      }
+
+      // Mask out writes to targets which aren't the pixel history color target
```

Comment:
I am wondering if the other target writes should be kept and not masked out. 
A test scenario I was thinking of was there are some draws to multiple render targets (RT0 and RT1). RT0 is the target of the pixel history. Later in the frame RT1 is bound as a texture source (or copied to an existing texture) for draws which output RT0 (and a different RT1).

If the output to RT1 is masked off then the data won't be present during replay for future events to use a texture source to output to RT0.

If possible be good to try that out with a simple test project (does not have to be RenderDoc test framework).

From a quick code review : D3D11, GL and Vulkan pixel history do not mask out the other render target writes i.e. from Vulkan code during the per fragment replay code
```
for(uint32_t i = 0; i < cbs->attachmentCount; i++)
{
  atts[i].blendEnable = 0;
  atts[i].colorWriteMask = VK_COLOR_COMPONENT_R_BIT | VK_COLOR_COMPONENT_G_BIT |
                            VK_COLOR_COMPONENT_B_BIT | VK_COLOR_COMPONENT_A_BIT;
}
```

Reply:
My understanding is that the actual replay is happening regardless of this pre-draw pass. I would expect that even if the callback were completely empty the replay would still be executed as described in the capture. Is this something I've misunderstood about renderdoc or the pixel history process?

Reply:
You are absolutely right